adal/cache_driver.py

```diff
@@ -164,11 +164,13 @@ def _refresh_entry_if_necessary(self, entry, is_resource_specific):
         now_plus_buffer = now + timedelta(minutes=Misc.CLOCK_BUFFER)
 
         if is_resource_specific and now_plus_buffer > expiry_date:
-            self._log.info('Cached token is expired.  Refreshing: %s', expiry_date)
-            return self._refresh_expired_entry(entry)
+            if TokenResponseFields.REFRESH_TOKEN in entry:
```

Comment:
No matched `else`  here means we return "None"? If yes, then it means the old expired entry will stay in the cache. My concern is, the subsequent token request will further push in a new valid token, but the old one might still get matched in future, rather the new token; so more and more new token will be added and the older ones still stay?  This will end up corrupting the token cache. If I get it right, I would suggest we delete this token before return None